src/Common/FileCacheFactory.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <Common/FileCache_fwd.h>
+
+#include <boost/noncopyable.hpp>
+#include <unordered_map>
+#include <mutex>
+
+namespace DB
+{
+
+/**
+ * Creates a FileCache object for cache_base_path.
+ */
+class FileCacheFactory final : private boost::noncopyable
+{
+public:
+    static FileCacheFactory & instance();
+
+    FileCachePtr getOrCreate(const std::string & cache_base_path, size_t max_size, size_t max_elements_size);
+
+private:
+    FileCachePtr getImpl(const std::string & cache_base_path, std::lock_guard<std::mutex> &);
```

Comment:
Unused parameter. Also, `std::lock_guard` as parameter looks suspicious.

Reply:
It's a style when you make sure that the method is called under a lock.

Reply:
Got it. That's nice